File: 백준/Silver/2578. 빙고/빙고.py
from sys import *
from collections import * 
from heapq import * 
from functools import *
from math import *
from itertools import *
import string

#setrecursionlimit(10**9)
input = stdin.readline

def check():
    global board
    
    count = 0
    
    for i in range(5):
        if board[i][0] == board[i][1] == board[i][2] == board[i][3] == board[i][4] == -1 : #가로 
            count += 1
        
        if board[0][i] == board[1][i] == board[2][i] == board[3][i] == board[4][i] == -1: # 세로
            count += 1
    
    
    
    if board[0][0] == board[1][1] == board[2][2] == board[3][3] == board[4][4] == -1: # 왼위 -> 오아래 대각
        count += 1
    
    if board[0][4] == board[1][3] == board[2][2] == board[3][1] == board[4][0] == -1: # 오위 -> 왼 아래 대각
        count += 1
    
    
   
    
    
        

    return count >= 3 


board = []
pos = [[0,0] for _ in range(26)]

for i in range(5):
    tmp = list(map(int,input().split()))
    board.append(tmp)
    for j in range(5):
        num = tmp[j]
        pos[num][0] , pos[num][1] = i , j 
    


find = False
ans = 0 
for i in range(5):
    
    numbers = list(map(int,input().split()))
    for j in range(5):
        count = (i) * 5  + (j+1) #  부른 횟수 

        x ,y = pos[numbers[j]] 
        board[x][y] = -1
        
        if not find:
            if check():
                ans = count
                find = True
                break


print(ans)
        
    
        

    
    
    

File: 백준/Silver/2578. 빙고/빙고.swift
var board : [Int:(Int,Int)] = [:]

var isOpen = Array(repeating:[false,false,false,false,false],count: 5)

for i in 0..<5 {
    let input = readLine()!.split{$0 == " "}.map{Int($0)!}
    
    for j in 0..<5 {
        board[input[j]] = (i,j)
    }
    
}

func diagonalLeftToRight() -> Bool {
    
    return isOpen[0][0] == true && isOpen[0][0] == isOpen[1][1] && isOpen[0][0] == isOpen[2][2] && isOpen[0][0] == isOpen[3][3] && isOpen[0][0] == isOpen[4][4]
}

func diagonaRightToLeft() -> Bool {
    
    return isOpen[0][4] == true && isOpen[0][4] == isOpen[1][3] && isOpen[0][4] == isOpen[2][2] && isOpen[0][4] == isOpen[3][1] && isOpen[0][4] == isOpen[4][0]
    
}

func horizon(_ r: Int) -> Bool {
    
    
    for i in 0..<5 {
        if isOpen[r][i] == false {
            return false
        }
    }
    
    return true
}

func vertical(_ c: Int) -> Bool {
    
    
    for i in 0..<5 {
        if isOpen[i][c] == false {
            return false
        }
    }
    
    return true
}

var count: Int = 1
var bingo: Set<[[Int]]> = .init()

var isBingo = false

for i in 0..<5 {
    let input = readLine()!.split{$0 == " "}.map{Int($0)!}
    
    for j in 0..<5 {
        
        let curr = board[input[j]]!
        
        let (r,c) = (curr.0,curr.1)
        
        isOpen[r][c] = true
        
        
        if diagonalLeftToRight() {
            bingo.insert([[0,0],[1,1],[2,2],[3,3],[4,4]])
        }
        
        if diagonaRightToLeft() {
            bingo.insert([[0,4],[1,3],[2,2],[3,1],[4,0]])
        }
        
        if horizon(r) {
            bingo.insert([[r,0],[r,1],[r,2],[r,3],[r,4]])
        }
        
        if vertical(c) {
            bingo.insert([[0,c],[1,c],[2,c],[3,c],[4,c]])
        }
        
        if bingo.count >= 3 {
            isBingo = true
            break
        }
        
        
        count += 1
    }
    
    if isBingo {
        break
    }
    
    
}

print(count)
